### PR TITLE
Reland "Text inline widgets, TextSpan rework"

### PR DESCRIPTION
**Pull request by GaryQian**
See the previous versions of this change: https://github.com/flutter/flutter/pull/30069 and https://github.com/flutter/flutter/pull/33794

This reland version includes minor performance fixes in `TextPainter.setPlaceholderDimensions()`. We now ignore placeholderDimensions when there are no placeholders in the textspan tree.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -4924,7 +4927,7 @@ class RichText extends LeafRenderObjectWidget {
   ///
   /// The [textDirection], if null, defaults to the ambient [Directionality],
   /// which in that case must not be null.
-  const RichText({
+  RichText({
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/basic.dart]: This is a breaking change

Comment (GaryQian) [reply packages/flutter/lib/src/widgets/basic.dart]: Yeah, this was announced already.

**Comment by jonahwilliams** [comment]
Removing the `const` from RichText is a breaking change, have you followed the breaking change procedure?

File: packages/flutter/lib/src/widgets/basic.dart (hunk 1 above)
Comment (GaryQian) [reply packages/flutter/lib/src/widgets/basic.dart]: Should be able to replace instances of RichText with Text.rich migrate.

**Comment by GaryQian** [comment]
Yes, breaking change procedure has been followed, about 2 weeks ago. #30069 has more history on this change.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 2]
```diff
@@ -465,6 +465,9 @@ class RenderParagraph extends RenderBox
   // specified.
   void _layoutChildren(BoxConstraints constraints) {
     RenderBox child = firstChild;
+    if (childCount == 0) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/paragraph.dart]: Why not do this as the first thing in the method?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/paragraph.dart (hunk 2 above)
Comment (GaryQian) [reply packages/flutter/lib/src/rendering/paragraph.dart]: Oops that's what I meant it to do.

**Pull request merged**